pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -615,7 +615,20 @@ func (e *Controller) checkLeftoverEndpoints() {
 			utilruntime.HandleError(fmt.Errorf("Unable to get key for endpoint %#v", ep))
 			continue
 		}
-		e.queue.Add(key)
```

Comment:
I think that we can just enqueue it with a certain delay

```
e.queue.AddAfter(key, 5*time.Second)
```

Reply:
Thank you very much for your review! Your solution looks more concise, but we may not be able to determine the specific value of the delay time very accurately. When the apiserver or etcd is under high load, the delay time may exceed the value we set, what do you think?

Reply:
oh wait, the controller waits for the informers to be synced before starting, it seems this scenario can not happen

https://github.com/kubernetes/kubernetes/blob/8e04f577d0b8a723c6c1f7986b228f532b6059ec/pkg/controller/endpoint/endpoints_controller.go#L198-L208

Reply:
Yes, but cache sync only guarantees that the data of the first list is consumed and stored in the informer, and cannot guarantee the synchronization sequence of newly added events. If a new service and endpoint are created just after the cache sync, this problem may still be triggered. This PR is to solve this scenario

Reply:
If the cache doesn't have the endpoint or service when the cleanup function function runs, wouldn't a subsequent event noting their creation would result in the Service being queued again?

Reply:
> If the cache doesn't have the endpoint or service when the cleanup function function runs, wouldn't a subsequent event noting their creation would result in the Service being queued again?

@robscott 
Yes, the service may be queued again, but if the service without a selector, syncService wouldn't recreate the deleted endpoint.
https://github.com/kubernetes/kubernetes/blob/8e04f577d0b8a723c6c1f7986b228f532b6059ec/pkg/controller/endpoint/endpoints_controller.go#L413-L417

Reply:
We should have a reproducer, this is evolving from the original description ...

Reply:
> We should have a reproducer, this is evolving from the original description ...

OK, I will add a test case to reproduce this problem

Reply:
I have updated the description, add a test case to reproduce this issue, PTAL. @aojea

Reply:
thanks for the quick response, I've tried the reproducer adding an additional debug line
```
diff --git a/pkg/controller/endpoint/endpoints_controller.go b/pkg/controller/endpoint/endpoints_controller.go
index 87c515f2f22..a65402d6613 100644
--- a/pkg/controller/endpoint/endpoints_controller.go
+++ b/pkg/controller/endpoint/endpoints_controller.go
@@ -601,6 +601,7 @@ func (e *Controller) checkLeftoverEndpoints() {
                utilruntime.HandleError(fmt.Errorf("Unable to list endpoints (%v); orphaned endpoints will not be cleaned up. (They're pretty harmless, but you can restart this component if you want another attempt made.)", err))
                return
        }
+       fmt.Println("DEBUG endpoints listed", len(list))
        for _, ep := range list {
                if _, ok := ep.Annotations[resourcelock.LeaderElectionRecordAnnotationKey]; ok {
                        // when there are multiple controller-manager instances,
```

I can see your point now, sorry for not able to get it before:
```
go test -v ./pkg/controller/endpoint -run TestCheckLeftoverEndpointsDeletedUnexpectedly -count 1
=== RUN   TestCheckLeftoverEndpointsDeletedUnexpectedly
I0802 12:19:43.681503 3929086 endpoints_controller.go:195] Starting endpoint controller
I0802 12:19:43.681561 3929086 shared_informer.go:240] Waiting for caches to sync for endpoint
I0802 12:19:43.782442 3929086 shared_informer.go:247] Caches are synced for endpoint 
DEBUG endpoints listed 1173
    endpoints_controller_test.go:444: svc count: 2000, endpoint count: 1899
--- FAIL: TestCheckLeftoverEndpointsDeletedUnexpectedly (0.14s)
FAIL
I0802 12:19:43.808651 3929086 endpoints_controller.go:212] Shutting down endpoint controller
FAIL    k8s.io/kubernetes/pkg/controller/endpoint       0.161s
FAIL
```

at the moment that `checkLeftoverEndpoints()` start running it has 1173 endpoint in the cache, we preloaded 1000, there are additional 1173, so I assume that this is the problem you are highlighting. However, that test is not realistic since the controller is the ones creating the endpoints, here you are the one creating custom Endpoints

Since you referenced services without selectors previusly, Is it possible that your problem is related to Services without selectors and custom Endpoints?

I mean, is `checkLeftoverEndpoints()` deleting your custom endpoints on restart?

Reply:
ok, then I think that we can change the approach, I personally prefer to avoid to do requests from the controller if is possible

@robscott the problem is that we don't have an `OwnerReferences` like in the endpointslice controller.
If we could have something to discriminate the "custom" endpoints  inside `checkLeftoverEndpoints()` we can just skip those endpoints and avoid the race ...  What about the annotation `EndpointsLastChangeTriggerTime`, can we use it? or is too hacky?

Reply:
I personally think that this API request is acceptable. It may only be requested when the conroller starts. If there is no orphaned endpoint or a new endpoint created when the controller starts, this API request will not be executed, and this request will hardly cause pressure on APIServer. 
Adding additional annotaion looks a bit hacky and not very friendly to the existing business. Do you have any other better suggestions to solve this problem? thanks.  @aojea @robscott

Reply:
discussion keeps going on mailing list https://groups.google.com/g/kubernetes-sig-network/c/ni73iB9ffjY/m/C15qKTOlAgAJ

I still think API request is too expensive for this scenario, and it doesn't solve the root cause (as Rob says in the thread)

Reply:
thanks for response, I will follow the discussion on the mailing list.